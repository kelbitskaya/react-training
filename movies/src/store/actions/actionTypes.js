export const FETCH_MOVIES_LIST_REQUEST = 'FETCH_MOVIES_LIST';
export const FETCH_MOVIES_LIST_SUCCESS = 'FETCH_MOVIES_LIST_SUCCESS';
export const FETCH_MOVIES_LIST_FAILURE = 'FETCH_MOVIES_LIST_FAILURE';
