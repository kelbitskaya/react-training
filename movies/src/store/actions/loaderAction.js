export default function loader(payload){
  return {
    type: 'LOADER',
    payload
  }
};

