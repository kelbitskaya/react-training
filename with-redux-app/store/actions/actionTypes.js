export const FETCH_MOVIES_LIST_REQUEST = 'FETCH_MOVIES_LIST';
export const FETCH_MOVIES_LIST_SUCCESS = 'FETCH_MOVIES_LIST_SUCCESS';
export const FETCH_MOVIES_LIST_FAILURE = 'FETCH_MOVIES_LIST_FAILURE';
export const CHANGE_MOVIES_LIST = 'CHANGE_MOVIES_LIST';
export const UPDATE_MOVIE = 'UPDATE_MOVIE';
export const UPDATE_SEARCH_TERM = 'UPDATE_SEARCH_TERM';
