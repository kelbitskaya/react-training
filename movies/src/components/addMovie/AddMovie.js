import React, { useState } from 'react';
import PropTypes from 'prop-types';
import {bindActionCreators} from "redux";
import {connect} from "react-redux";
import Input from '../common/input/Input';
import Button from '../common/button/Button';
import MultiSelectInput from '../common/multiSelectInput/MultiSelectInput';
import Datepicker from '../common/datepicker/Datepicker';
import CloseButton from '../closeButton/CloseButton';
import MovieActionPopup from '../movieActionPopup/MovieActionPopup';
import Constants from '../constants';
import {editMoviesList} from "../../store/actions/actions";

const matchDispatchToProps = (dispatch) => {
  return bindActionCreators({updateMovies: editMoviesList}, dispatch)
};

const mapStateToProps = (state) => {
  return {
    state,
  }
};

const AddMovie = (props) => {
  const {state, updateMovies} = props;
  const [currentMovies, setCurrentMovies] = useState(state.movies);
  const [isMovieAction, isMovieActionPopup] = useState(false);
  const { handleClose } = props;

  const handleSubmit = (e) => {
    e.preventDefault();

    const formData = new FormData(e.target);
    const values = Object.fromEntries(formData.entries());

    const newList = [...state.movies.data, values];
    state.movies.data = newList;
    setCurrentMovies( state.movies);

    console.log('v5', values);
    updateMovies(currentMovies);
    isMovieActionPopup(true);
  };

  return (
    <>
      <CloseButton handleClose={handleClose} />
      <form onSubmit={handleSubmit}>
        <p className="modal-title">Add Movie</p>
        <Input
          id="movie-title"
          title="Title"
          labelClassName="input-label"
          name="title"
        />
        <Datepicker
          id="movie-release"
          title="Release date"
          labelClassName="input-label"
          value={this}
          name="release_date"
        />
        <Input
          id="movie-url"
          title="Movie url"
          labelClassName="input-label"
          name="poster_path"
        />
        <div className="modal-dropdown">
          <span className="modal-label-text">Genre</span>
          <MultiSelectInput options={Constants.GENRE} />
        </div>
        <Input
          id="movie-overview"
          title="overview"
          labelClassName="input-label"
          name="overview"
        />
        <Input
          id="movie-runtime"
          title="runtime"
          labelClassName="input-label"
          name="runtime"
        />
        <div className="add-movie-buttons">
          <Button
            className="button button_secondary"
            title="RESET"
            handleClick={() => {}}
          />
          <Button
            className="button button_primary"
            title="SUBMIT"
            type="submit"
          />
        </div>
      </form>
      <MovieActionPopup
        isOpen={isMovieAction}
        title="CONGRATULATION!"
        message="The movie has been added successfully!"
        handleClose={handleClose}
      />
    </>
  );
}

AddMovie.propTypes = {
  handleClose: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, matchDispatchToProps)(AddMovie);
