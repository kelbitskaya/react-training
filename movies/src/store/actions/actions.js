import axios from "axios";
import * as types from './actionTypes';

export const fetchMoviesRequest = () => {
  return {
    type: types.FETCH_MOVIES_LIST_REQUEST
  }
};

export const fetchMoviesSuccess = movies => {
  return {
    type: types.FETCH_MOVIES_LIST_SUCCESS,
    payload: movies
  }
};

export const fetchMoviesFailure = error => {
  return {
    type: types.FETCH_MOVIES_LIST_FAILURE,
    payload: error
  }
};

export const fetchMovies = (sortBy, filter, title) => {
  return function (dispatch) {
    dispatch(fetchMoviesRequest());

    axios.get('http://localhost:4000/movies', {
      params: {
        limit: 9,
        sortOrder: 'asc',
        sortBy: sortBy || 'release_date',
        filter: filter,
        search: title,
        searchBy: 'title'
      }
    })
      .then(response => {
        dispatch(fetchMoviesSuccess(response.data));
      })
      .catch(error => {
        dispatch(fetchMoviesFailure(error.message))
      })
  }
};

export const deleteMovie = (id, searchType) => async dispatch => {
  try {
    await axios.delete(`http://localhost:4000/movies/${id}`);
    dispatch(fetchMovies('', '', searchType));
  } catch (error) {
    console.error();
  }
};

export const updateMovieInList = data => {
  return {
    type: types.UPDATE_MOVIE,
    payload: data
  }
};

export const getMovieById = (id) => async dispatch => {
  await axios({
    method: 'get',
    url: `http://localhost:4000/movies/${id}`,
  }).then(response => {
    dispatch(updateMovieInList(response.data));
  });
};



export const updateMovie = (data, searchType, sortBy) => async dispatch => {
  await axios({
    method: 'put',
    url: 'http://localhost:4000/movies/',
    headers: {'Content-Type': 'application/json'},
    data: JSON.stringify({...data})
  });
  dispatch(fetchMovies(sortBy, '', searchType));
};

export const addMovie = (data) => async dispatch => {
  await axios({
    method: 'post',
    url: 'http://localhost:4000/movies/',
    headers: {'Content-Type': 'application/json'},
    data: data
  });
  dispatch(fetchMovies('', '', ''));
};

export const editMoviesList = movies => {
  return {
    type: types.CHANGE_MOVIES_LIST,
    payload: movies
  }
};

