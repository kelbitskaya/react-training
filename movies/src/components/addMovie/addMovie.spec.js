import React from "react";
import AddMovie from "./AddMovie";
import rerender from "react-test-renderer";
import { Provider } from "react-redux";
import {addMovie} from "../../store/actions/actions";
import configureStore from "redux-mock-store";

const props = {
  handleClose: ()=>{},
  updateMovies: jest.fn(),
  isSubmitting: ()=>{},
  handleChange: ()=>{},
  handleBlur: ()=>{},
  handleSubmit: ()=>{},

  state: {
    movies: {
      data: [],
    }
  },
  values: {
    title: '',
    poster_path: '',
    overview: '',
    runtime: '',
  },
  errors: {
    title: '',
    poster_path: '',
    overview: '',
    runtime: '',
    release_date: '',
  },
};

describe("AddMovie", () => {

  const handleClose = jest.fn();

  test("AddMovie snapshot", () => {
    const initialStore = {};
    const store = configureStore([])(initialStore);
    store.dispatch = jest.fn();
    const onSubmit = store.dispatch((values) => addMovie(values));
    const component = rerender.create(
      <Provider store={store}>
        <AddMovie
                {...props}
                handleClose={handleClose}
                onSubmit={onSubmit} />
      </Provider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
