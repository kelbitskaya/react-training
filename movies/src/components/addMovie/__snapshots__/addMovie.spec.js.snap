// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovie AddMovie snapshot 1`] = `
Array [
  <button
    aria-label="Close"
    className="button-close"
    onClick={[MockFunction]}
    type="button"
  />,
  <p
    className="modal-title"
  >
    Add Movie
  </p>,
  <form
    onSubmit={[Function]}
  >
    <label
      className="input-label"
      htmlFor="title"
    >
      <span
        className="modal-label-text"
      >
        Title
      </span>
      <input
        className="input"
        id="title"
        name="title"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter movie title"
        type="text"
        value=""
      />
    </label>
    <label
      className="input-label"
      htmlFor="release_date"
    >
      <span
        className="modal-label-text"
      >
        RELEASE DATE
      </span>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            className="input myClass"
            disabled={false}
            name="release_date"
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={false}
            type="text"
            value=""
          />
        </div>
      </div>
    </label>
    <label
      className="input-label"
      htmlFor="poster_path"
    >
      <span
        className="modal-label-text"
      >
        MOVIE URL
      </span>
      <input
        className="input"
        id="poster_path"
        name="poster_path"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter movie poster path"
        type="text"
        value=""
      />
    </label>
    <div
      className="modal-dropdown"
    >
      <span
        className="modal-label-text"
      >
        Genre
      </span>
      <div
        className="go2139093995 multi-select"
      >
        <div
          aria-expanded={false}
          aria-readonly={true}
          className="go2646822163 dropdown-container"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={0}
        >
          <div
            className="go2642161244 dropdown-heading"
            onClick={[Function]}
          >
            <div
              className="dropdown-heading-value"
            >
              <span
                className="gray"
              >
                Select...
              </span>
            </div>
            <svg
              className="dropdown-heading-dropdown-arrow gray"
              fill="none"
              height="24"
              stroke="currentColor"
              strokeWidth="2"
              width="24"
            >
              <path
                d="M6 9L12 15 18 9"
              />
            </svg>
          </div>
        </div>
      </div>
      <input
        name="genres"
        type="hidden"
        value=""
      />
    </div>
    <label
      className="input-label"
      htmlFor="overview"
    >
      <span
        className="modal-label-text"
      >
        OVERVIEW
      </span>
      <input
        className="input"
        id="overview"
        name="overview"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter movie overview"
        type="text"
        value=""
      />
    </label>
    <label
      className="input-label"
      htmlFor="runtime"
    >
      <span
        className="modal-label-text"
      >
        RUNTIME
      </span>
      <input
        className="input"
        id="runtime"
        name="runtime"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter movie runtime"
        type="text"
        value=""
      />
    </label>
    <div
      className="add-movie-buttons"
    >
      <button
        className="button button_primary"
        onClick={[MockFunction]}
        type="button"
      >
        RESET
      </button>
      <button
        className="button button_primary"
        disabled={false}
        type="submit"
      >
        SUBMIT
      </button>
    </div>
  </form>,
]
`;
